File: solutions/haskell/clock/1/src/Clock.hs
module Clock (addDelta, fromHourMin, toString) where

data Clock = Clock Int Int
  deriving (Eq)

fromHourMin :: Int -> Int -> Clock
fromHourMin a b = Clock (mod (a + div b 60) 24) (mod b 60)

toString :: Clock -> String
toString (Clock a b) = s (show a) ++ ":" ++ s (show b)

addDelta :: Int -> Int -> Clock -> Clock
addDelta hour min (Clock a b) = fromHourMin (hour + a) (min + b)

s :: String -> String
s "" = "00"
s [x] = '0' : [x]
s x = x

File: solutions/kotlin/clock/1/src/main/kotlin/Clock.kt
class Clock(h: Int, m: Int) {
    //TODO: implement proper constructor
    var min = m.mod(60)
    var hour = (h + (m - m.mod(60)).div(60)).mod(24)

    //TODO: find a convenient way to take over `toString()` and `equals()`
    override fun toString(): String {
        return "%02d:%02d".format(hour, min)
    }

    override fun equals(other: Any?): Boolean {
        return other is Clock && this.min == other.min && this.hour == other.hour
    }

    override fun hashCode(): Int {
        return min + 60 * hour
    }

    fun subtract(minutes: Int) {
        val m = min - minutes
        val h = hour
        min = m.mod(60)
        hour = (h + (m - m.mod(60)).div(60)).mod(24)
    }

    fun add(minutes: Int) {
        
        val m = min + minutes
        val h = hour
        min = m.mod(60)
        hour = (h + (m - m.mod(60)).div(60)).mod(24)
    }
}
